src/rgw/s3select.h

```diff
@@ -0,0 +1,852 @@
+#ifndef __S3SELECT__
+#define __S3SELECT__
+
+#include <boost/spirit/include/classic_core.hpp>
+#include <iostream>
+#include <string>
+#include <list>
+#include "s3select_oper.h"
+#include <boost/function.hpp>
+#include <boost/bind.hpp>
+#include <functional>
+
+using namespace std;
+using namespace BOOST_SPIRIT_CLASSIC_NS;
+#define _DEBUG_TERM {string  token(a,b);std::cout << __FUNCTION__ << token << std::endl;}
+
+/// AST builder
+
+class s3select_projections {
+
+    private:
+        list<base_statement *> m_projections;
```

Comment:
is there a chance this could be a std::vector or, given how easy it actually is to use, maybe use boost::intrusive::list (potentially most efficient)?

Reply:
the AST builder is running once per query life time , there is no performance issues on those parts.
the use of container (STL list in this case)  is for building the execution tree , once its created (AST) , the input data(s3 object) is bind to it. 
the execution tree is analog to binary program running on different inputs.

Reply:
the point is std::list is I'm sure not a bottleneck, allocating twice where vector would allocated once--vector is good esp when you can predict the number of elements (can reserve());  intrusive list is almost as easy to use and doesn't allocate

Reply:
ok, will use intrusive list.
stl-list is used because of push-pop semantics; can not predict number of elements since it depends on query length and its complexity.